projects/core/directives/dropdown/dropdown.component.ts

```diff
@@ -0,0 +1,120 @@
+import {TUI_ANIMATION_OPTIONS} from '@taiga-ui/core/tokens';
+import {TuiPoint} from '@taiga-ui/core/types';
+import {Observable} from 'rxjs';
+import {takeUntil} from 'rxjs/operators';
+
+import {TuiDropdownDirective} from './dropdown.directive';
+import {TuiDropdownHoverDirective} from './dropdown-hover.directive';
+import {TUI_DROPDOWN_OPTIONS, TuiDropdownOptions} from './dropdown-options.directive';
+
+/**
+ *  This component is used to show template in a portal using default style of white rounded box with a shadow
+ */
+// @bad TODO: OnPush
+@Component({
+    selector: `tui-dropdown`,
+    templateUrl: `./dropdown.template.html`,
+    styleUrls: [`./dropdown.style.less`],
+    changeDetection: ChangeDetectionStrategy.Default,
+    providers: [TuiDestroyService, TuiPositionService],
+    animations: [tuiDropdownAnimation],
+})
+export class TuiDropdownComponent {
+    @HostBinding(`@tuiDropdownAnimation`)
+    readonly dropdownAnimation = {
+        value: TuiDropdownAnimation.FadeInTop,
+        ...this.animationOptions,
+    };
+
+    constructor(
+        @Inject(TuiPositionService) position$: Observable<TuiPoint>,
+        @Inject(TuiDestroyService) destroy$: Observable<void>,
+        @Inject(TuiDropdownDirective) readonly directive: TuiDropdownDirective,
+        @Inject(ElementRef) private readonly elementRef: ElementRef<HTMLElement>,
+        @Inject(TuiRectAccessor) private readonly accessor: TuiRectAccessor,
+        @Inject(WINDOW) private readonly windowRef: Window,
+        @Inject(TUI_ANIMATION_OPTIONS)
+        private readonly animationOptions: AnimationOptions,
+        @Inject(TUI_DROPDOWN_OPTIONS) private readonly options: TuiDropdownOptions,
+        @Optional()
+        @Inject(TuiDropdownHoverDirective)
+        private readonly hoverDirective: TuiDropdownHoverDirective | null,
+    ) {
+        position$.pipe(takeUntil(destroy$)).subscribe(([top, left]) => {
+            this.update(top, left);
+        });
+    }
+
+    onHoveredChange(hovered: boolean): void {
+        if (this.hoverDirective) {
+            this.hoverDirective.toggle(hovered);
+        }
+    }
+
+    onTopFocus(): void {
+        this.moveFocusOutside(true);
+    }
+
+    onBottomFocus(): void {
+        this.moveFocusOutside(false);
+    }
+
+    private update(top: number, left: number): void {
+        const {style} = this.elementRef.nativeElement;
+        const {right} = this.elementRef.nativeElement.getBoundingClientRect();
+        const {limitWidth, maxHeight} = this.options;
+        const {innerHeight, scrollX, scrollY} = this.windowRef;
+        const rect = this.accessor.getClientRect();
+        const offsetX = this.directive.position === `fixed` ? 0 : scrollX;
+        const offsetY = this.directive.position === `fixed` ? 0 : scrollY;
+
+        top = top + offsetY;
+        left = left + offsetX;
+
+        const intersect = left < rect.right && right > rect.left && top < offsetY + 12;
```

Comment:
what is `12` / `8` number?

can we put them into naming constants?

Reply:
4 is the offset around host, 8 is the offset around screen edges. Added a constant.